newlib: do not attempt to overallocate heap

yroeht opened the pull request:
In malloc_prepare, in the case where the configured Newlib max heap size is greater than the physically available memory, we allocate as much as possible. Because k_mem_free_get returns all the available page frames but k_mem_map reserves two guard pages (one at the start, one at the end), to avoid a failure in k_mem_map that results in no heap being allocated, subtract the size of those two guard pages.

Signed-off-by: Théophile Ranquet <theophile.ranquet@gmail.com>

yroeht committed:
newlib: do not attempt to overallocate heap

In malloc_prepare, in the case where the configured Newlib max heap size
is greater than the physically available memory, we allocate as much as
possible. Because k_mem_free_get returns all the available page frames
but k_mem_map reserves two guard pages (one at the start, one at the
end), to avoid a failure in k_mem_map that results in no heap being
allocated, subtract the size of those two guard pages.

Signed-off-by: Théophile Ranquet <theophile.ranquet@gmail.com>

keith-packard commented on lib/libc/newlib/libc-hooks.c:
```diff
@@ -104,8 +104,13 @@ static int malloc_prepare(const struct device *unused)
 
 #ifdef USE_MALLOC_PREPARE
 #ifdef CONFIG_MMU
+	/* Need extra for the guard pages (before and after) which we
```
Is this always true? Or does this depend on some other configuration which actually places guard pages around allocations?

cfriedt reviewed (changes requested):
Can we have a couple of test cases please?

yroeht committed:
mmu: report free virtual memory

There is currenty a facility to obtain the amount of free physical
memory, but a way to obtain free virtual memory is also needed to know
how much we can possibly allocate. This is used by newlib in
malloc_prepare.

Signed-off-by: Théophile Ranquet <theophile.ranquet@gmail.com>

yroeht commented:
Twister failure seems unrelated?

```
FAILED: zephyr/CMakeFiles/zephyr.dir/subsys/usb/device/class/cdc_acm.c.obj 
ccache /opt/toolchains/zephyr-sdk-0.15.2/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DKERNEL -DNRF52840_XXAA -DTC_RUNID=4ba0f2786316c35d581fa2b8fff85624 -D__PROGRAM_START -D__ZEPHYR__=1 -I../../../../../../../kernel/include -I../../../../../../../arch/arm/include -I../../../../../../../include -Izephyr/include/generated -I../../../../../../../soc/arm/nordic_nrf/nrf52 -I../../../../../../../soc/arm/nordic_nrf/common/. -I../../../../../../../subsys/usb/device -I../../../../../../../subsys/testsuite/include -I../../../../../../../subsys/testsuite/ztest/include -I/__w/zephyr/modules/hal/cmsis/CMSIS/Core/Include -I/__w/zephyr/modules/hal/nordic/nrfx -I/__w/zephyr/modules/hal/nordic/nrfx/drivers/include -I/__w/zephyr/modules/hal/nordic/nrfx/mdk -I../../../../../../../modules/hal_nordic/nrfx/. -isystem ../../../../../../../lib/libc/minimal/include -isystem /opt/toolchains/zephyr-sdk-0.15.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.1.0/include -isystem /opt/toolchains/zephyr-sdk-0.15.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.1.0/include-fixed -fno-strict-aliasing -Os -imacros /__w/zephyr/zephyr/twister-out/bl654_usb/zephyr/tests/kernel/common/kernel.common/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfp16-format=ieee --sysroot=/opt/toolchains/zephyr-sdk-0.15.2/arm-zephyr-eabi/arm-zephyr-eabi -imacros /__w/zephyr/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -Werror -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/__w/zephyr/zephyr/tests/kernel/common=CMAKE_SOURCE_DIR -fmacro-prefix-map=/__w/zephyr/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/__w/zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT zephyr/CMakeFiles/zephyr.dir/subsys/usb/device/class/cdc_acm.c.obj -MF zephyr/CMakeFiles/zephyr.dir/subsys/usb/device/class/cdc_acm.c.obj.d -o zephyr/CMakeFiles/zephyr.dir/subsys/usb/device/class/cdc_acm.c.obj -c /__w/zephyr/zephyr/subsys/usb/device/class/cdc_acm.c
/__w/zephyr/zephyr/subsys/usb/device/class/cdc_acm.c:63:2: error: #warning "USB_CDC_ACM_LOG_LEVEL forced to LOG_LEVEL_NONE" [-Werror=cpp]
   63 | #warning "USB_CDC_ACM_LOG_LEVEL forced to LOG_LEVEL_NONE"
      |  ^~~~~~~
```

cfriedt commented:
@yroeht - fixed by #53419

yroeht commented:
> @yroeht - fixed by #53419.

Is this a Zephyr pull/issue? Could not find it.

> Can you rebase?

Rebased on origin/main (see below), no change.

```
commit 9cf89603dfffd320d9b3d127e5cccc1d8c40b24e
Author: Serhiy Katsyuba <serhiy.katsyuba@intel.com>
Date:   Wed Dec 14 16:05:45 2022 +0100

    intel_adsp: dai: Add support for ALH streams 2 and 3

```

dcpleung commented on include/zephyr/sys/mem_manage.h:
```diff
@@ -290,6 +290,13 @@ void z_phys_unmap(uint8_t *virt, size_t size);
+ *
+ * @return Free virtual RAM, in bytes
+ */
+int k_virt_free_range_max_get(void);
```
Maybe `k_mem_virt_free_get()`? And this should return `size_t`.

The name used here somehow seems like it would return the range of memory.

dcpleung commented on kernel/mmu.c:
```diff
@@ -932,6 +932,28 @@ void z_mem_manage_boot_finish(void)
+			++free_pages;
+			continue;
+		}
+		free_pages_max = free_pages > free_pages_max ? free_pages : free_pages_max;
+		free_pages = 0;
```
It seems like you are trying to return the size of the largest continuous block of virtual memory. If so, the function description in the header file needs to be clarified. And the name of the function needs to convey this too.

yroeht replied:
Ok I will use k_mem_virt_free_get() as you have suggested

cfriedt reviewed (changes requested):
You're adding both API and fixing a bug in this PR.

Please add tests to
* demonstrate nominal functionality of the new API
* demonstrate the bug (before the bugfix commit is merged)
* demonstrate that the bugfix fixes the bug (after the bugfix commit is merged)
